Add Connection Pool Object

philsbln opened the pull request:
This PR adds a first proposal for Connection Pools to TAPS as discussed in Issue #266.

**This is an alternative to PR #298.**

Pooled Connections are only useful for request/response protocols like HTTP and serve the following purposes:
 - Provide a clean abstraction for HTTP/3 that is backward compatible to HTTP/0.9
 - Remove the motivation to map QUIC streams to Messages (by removing the drawback of complication HTTP/3 mapping when they are mapped to TAPS connections)
 - Allow Per-Request Path- and Endpoint Selection

See [my post on the mailing list](https://mailarchive.ietf.org/arch/msg/taps/u4imn1IKFBeSxwJRJK_HNwEoT98) for a more detailed description of this topic.

britram commented on draft-ietf-taps-interface.md:
```diff
@@ -997,17 +1052,22 @@ and takes optional per-Message properties (see {{send-basic}}). All Send actions
 ~~~
-Connection.Send(messageData, messageContext?, endOfMessage?)
+msgRef := Connection.Send(messageData, messageContext?, endOfMessage?)
+msgRef := ConnectionPool.Send(reqRef?, messageData, messageContext?, endOfMessage?)
```
Exposing the distinction between a Connection and a ConnectionPool in this way seems to add complexity to the API for no benefit. Why does the application need to know a connection pool might be in use? Isn't one of the points of TAPS to allow multiplexing and pooling when possible but to provide nominal functionality when not?

The request reference is indeed useful here (if we accept that TAPS should be useful to build interfaces atop request/response semantics, which I do think is the case) to tell the TAPS system that one message should use the same low-level transport (Transport Stack Instance in Post terminology, IIRC) as another, but ISTM this could be added to every Received event and Send call as an optional parameter.

tfpauly commented:
Thanks for writing this up!

I agree very much with @britram that the notion of a separate object for a ConnectionPool does not seem to add direct value, but makes the interface more complex.

One of the key notions of the architecture is that the "Connection" is not necessarily analogous to a TCP connection, and I think that really makes the Connection abstraction able to do the same thing you're trying to add with a "Pool". For example, in HTTP/2, a Connection will be just one stream, that's multiplexed. And since each stream is effectively a single HTTP request/response, that means that we're really just treating a Connection as an HTTP request/response pair. So, over HTTP/1.1, or the like, a Connection for a request/response pair ends up "pooling" onto existing TLS/TCP connections. This can and should be automatic, perhaps with the ability to define boundaries of when not to automatically group.

philsbln commented:
I just updated the Connection Pool objects to be more Connection like and added the ```reqRef``` as parameter to the normal send and receive. So a connection pool really works like a connection and only differs in creation by calling ```RequestorPool()``` instead of ```Initiate()``` and ```ResponderPool()``` instead of ```Listen()```.

philsbln commented:
Closing this based on the result of the IETF-104 discussion
